Deprecate BOOST distributions

bwspenc opened the pull request:
ref #12688

Point the deprecation messages for all the Boost distributions to
the appropriate non-Boost distribution.

I had to move the implmentations for some of the Distribution classes
into a separate class, and add a wrapper Distribution class because of
difficulties in using both registerMooseObjectReplaced() and
registerMooseObjectAliased() together. This gets a head start on
moving all of these over to classes without Distribution at the end
of the name anyway.

bwspenc committed:
Deprecate BOOST distributions ref #12688

Point the deprecation messages for all the Boost distributions to
the appropriate non-Boost distribution.

I had to move the implmentations for some of the Distribution classes
into a separate class, and add a wrapper Distribution class because of
difficulties in using both registerMooseObjectReplaced() and
registerMooseObjectAliased() together. This gets a head start on
moving all of these over to classes without Distribution at the end
of the name anyway.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/519758/) on adc9ef3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15226/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15226/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c5f90a85163476f6d5e671dfb0ded56effc378e0`

aeslaughter commented:
A @zachmprince can we make this your 2nd of 5 reviews?

zachmprince commented on modules/stochastic_tools/src/distributions/BoostLognormalDistribution.C:
```diff
@@ -9,27 +9,18 @@
 {
-  _distribution_unique_ptr = libmesh_make_unique<boost::math::lognormal_distribution<Real>>(
-      getParam<Real>("location"), getParam<Real>("scale"));
 }
```
Is there a reason we are keeping these functions around? It seems to clutter up our distributions folder.

bwspenc replied:
We were previously using some trickery with `registerMooseObjectReplaced` to keep the old class and register the new name pointing to the old class. That caused me a lot of trouble when I was trying to use `registerMooseObjectReplaced` on the Boost distributions to point to a class that didn't exist (the one without `Distribution` at the end of the name).

I thought it was easiest to just move the code to the classes without `Distribution` in the name (because those are the ones that we are going to keep, and then make classes with `Distribution` at the end that are simply wrappers around the new classes that make them available with the deprecated names. When all these classes are deprecated on 7/1/2020, we just need to go through and remove all classes with `Distribution` at the end of them and all the `Boost` classes. It does end up with some extra files in the directory, but it's temporary.

zachmprince replied:
Oh okay, I guess there is no easy way to replace and alias at the same time. I have 7/1/2020 marked on my calendar now.

zachmprince commented on modules/stochastic_tools/src/distributions/BoostNormal.C:
```diff
@@ -0,0 +1,37 @@
+// registerMooseObjectReplaced("StochasticToolsApp",
+//                            BoostNormal,
+//                            "07/01/2020 00:00",
+//                            Normal);
```
Left over comments?

bwspenc replied:
Whoops!  I should have all these cleaned up now.

zachmprince commented on modules/stochastic_tools/src/distributions/BoostWeibull.C:
```diff
@@ -0,0 +1,37 @@
+// registerMooseObjectReplaced("StochasticToolsApp",
+//                            BoostWeibull,
+//                            "07/01/2020 00:00",
+//                            Weibull);
```
Left over

zachmprince commented on modules/stochastic_tools/src/distributions/Weibull.C:
```diff
@@ -0,0 +1,87 @@
+registerMooseObject("StochasticToolsApp", Weibull);
+registerMooseObjectAliased("StochasticToolsApp", Weibull, "WeibullDistribution");
+// registerMooseObjectReplaced("StochasticToolsApp", WeibullDistribution, "07/01/2020 00:00",
+// Weibull);
```
Left over

zachmprince reviewed (commented):
Two major comments:
1. I don't think we should be deprecating our boost distributions.
1. I don't know why the *Distribution classes are being kept around

bwspenc committed:
Copy markdown files for distribution objects with new names ref #12688

Now that I changed how we register these classes, we need md files
corresponding to them.

bwspenc committed:
Rename and make wrappers for remaining deprecated distributions ref #12688

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/520426/) on b635d97 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15226/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15226/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3447b9ff576548170175de75811207e71130cf19`

bwspenc commented on modules/stochastic_tools/src/distributions/BoostLognormal.C:
```diff
@@ -0,0 +1,32 @@
+
+#include "libmesh/auto_ptr.h"
+
+registerMooseObjectReplaced("StochasticToolsApp", BoostLognormal, "07/01/2020 00:00", Lognormal);
```
The main motivation for removing the Boost distributions is that it's giving us a lot of headaches with our testing to test things both with and without Boost. The Grizzly build was broken for the last month because we were trying to figure out how to suppress tons of debug warnings that were coming from Boost. We've wasted weeks of time debugging Boost-related issues now, for what seems to me like little benefit.

I haven't done any performance benchmarking between our native distributions and the Boost ones -- I guess I assumed that the cost of evaluating distributions would be in the noise compared to everything else we do, but maybe that's not always the case. I'm curious just how much slower they are. What's the easiest way to see?  Just generate a bunch of Monte Carlo samples using a Normal distribution?

bwspenc replied:
My guess is that the main performance bottleneck is in `Normal::quantile()`, which is probably the trickiest of all of these functions, and I'm sure is implemented numerically in a way that differs from how Boost does it. Multiple distributions call this from their `quantile()` methods. Maybe we can do a little research in how to speed this up.

bwspenc replied:
I just looked at the Boost code, and it's calling a function to compute the inverse error function. Our current implementation uses a rational approximation that avoids doing that, but maybe it's slower. We are raising the same number to the same power many times in that function, which can be improved upon. I'm going to take a crack at that in another commit on this PR.

zachmprince replied:
Ah, I see why Boost created such a headache. I will try some performance testing between the two objects and see the difference because of `quantile`.

Evaluating the distribution could become a bottleneck when we are producing a bunch of samples and evaluating a fast surrogate model. But if we see only like a 10% speedup because of boost, it still might not be worth keeping boost around and our own functions are fast enough.

bwspenc replied:
OK, I just did some performance testing.  I modified one of the tests in `stochastic_tools/test/tests/samplers/monte_carlo` to generate 1e6 MC samples using a normal distribution (which exercises its `quantile()` function. If I turn off output, so that I don't count I/O time, it runs in 0.825s. I didn't compare to the Boost version because it's such a pain to re-build with Boost, but I compared it with the Uniform distribution, the quantile for which is basically free, and I get a run time of 0.774s. I have a hard time believing that if there is any increased performance in the Boost version that it will matter.

zachmprince replied:
Okay I was able to do a test between Normal with your improvements and BoostNormal. Looks like your implementation is faster than boost!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/520507/) on 224b63c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15226/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15226/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3447b9ff576548170175de75811207e71130cf19`

bwspenc committed:
Improve performance of Normal::quantile() ref #12688

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/520591/) on 8df4537 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15226/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15226/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3447b9ff576548170175de75811207e71130cf19`

zachmprince reviewed (commented):
I concede, and the boost distributions can be deprecated.

I still think a lot of the \*Distribution classes can be removed by using `registerMooseObjectRenamed`. But, the Boost\*Distribution classes will probably need to stay.

For the documentation though, we can use the `remove.yml` file to ignore the deprecated objects. Then we can remove all the duplicate documentation.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/520644/) on e2a991a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15226/site)

This comment will be updated on new commits.

bwspenc commented on modules/stochastic_tools/src/distributions/JohnsonSBDistribution.C:
```diff
@@ -8,10 +8,7 @@
-registerMooseObjectAliased("StochasticToolsApp", JohnsonSBDistribution, "JohnsonSB");
 registerMooseObjectReplaced("StochasticToolsApp",
                             JohnsonSBDistribution,
                             "07/01/2020 00:00",
```
Oh, wow, you're right!  If I knew about `registerMooseObjectRenamed()`, that would have saved me a bunch of trouble.  Yes, I'll do this and get rid of all the files with `Distribution` at the end of their name. I won't need the separate markdown for them either. All I did was copy the existing markdown. You have good suggestions on how to improve it, and I'll try do do as many of those as I can, but really, I was just moving stuff around.

bwspenc committed:
Fix up documentation to address comments ref #12688

bwspenc commented:
@zachmprince We've fixed the issue in Grizzly and I've made the changes as we discussed to this PR. This adds the second set of classes, which I think we should do to prevent breakage of Grizzly, which is creating the old `Deprecated` classes currently. I think this is ready to go on my side unless you have other comments, so if you could look this through when you get a chance again, that would be great.

Once this goes through and I change Grizzly to use the classes without the `Deprecated` postfix, we could eliminate some of these classes through the use of aliased registration, but I don't know that it's really worth the trouble since we're just going to remove them all in a month and a half anyway.

zachmprince commented:
@bwspenc the code looks good right now. Once Grizzly is fixed, I think we can immediately delete all the *Distribution files and use the renamed deprecation. I would still like the documentation a little cleaner in the meantime. I will add a commit to your branch so that we can remove all the *Distribution.md files. That way, they won't show up in the Distributions list.

zachmprince committed:
Removing deprecated distributions from documentation #12688

zachmprince commented:
@bwspenc I was able to delete all the deprecated distributions from the documentation and created a pull request to your branch.

bwspenc committed:
Merge pull request #21 from zachmprince/deprecate_boost_distributions

Removing deprecated distributions from documentation

bwspenc commented:
OK, thanks, @zachmprince I just merged your PR into my branch. Those changes all looked fine to me. Assuming the tests pass, I think this is ready to go.

aeslaughter commented on modules/stochastic_tools/doc/content/source/distributions/BoostLognormal.md:
```diff
@@ -0,0 +1,38 @@
+The following input file defines a lognormal distribution with the location parameter -0.371 and the
+scale parameter 0.52.
+
+!listing modules/stochastic_tools/test/tests/distributions/boost_lognormal.i block=Distributions
```
I would just punt on adding images for now, but I do like the idea.

aeslaughter commented:
@bwspenc I put a PR into BISON that should fix the documentation failure that is showing up because of this work.

moosebuild commented:
Job [Application Documentation](https://civet.inl.gov/job/522891/) on 5245c36 : invalidated by @bwspenc

BISON PR has been merged, which should fix this failure

zachmprince commented:
Has @aeslaughter 's PR been merged into bison master yet?

aeslaughter commented:
No. That will take a few more hours. It is in valgrind testing, which takes
about 4 hours.

On Tue, May 12, 2020 at 2:13 PM Zachary Prince <notifications@github.com>
wrote:

> Has @aeslaughter <https://github.com/aeslaughter> 's PR been merged into
> bison master yet?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15226#issuecomment-627568618>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACTKVPD2UP5VL6YBHQBTYLRRGUXPANCNFSM4MZ5WVTA>
> .
>

moosebuild commented:
Job [Application Documentation](https://civet.inl.gov/job/522891/) on 5245c36 : invalidated by @bwspenc

bwspenc commented:
The BISON change made it through and after I invalidated the test, the tests pass.  Since it's been approved by @aeslaughter, I'm going ahead and merging this.
